### PR TITLE
Hold Taps

### PR DESCRIPTION
**Pull request by okke-formsma**
This merge implements the behavior discussed in https://github.com/zmkfirmware/zmk/issues/117.

The tap-hold behavior will soak new position-events until it can make a decision to 'hold' or 'tap', after which all queued keycodes are bubbled. 

We get three "flavors" of mod-taps: "hold-preferred" (default), "balanced" and "tap-preferred". 

* The hold-preferred is default as it gives the fastest response; as soon as another key is pressed, the hold functionality is enabled. (Or if the timer runs out)
* The balanced behavior decides on hold if another key is pressed and released while the tap-hold was active (Or if the timer runs out)
* The tap-preferred behavior only decides on hold if the timer runs out.

Not all behaviors from the PDF are implemented; specifically 3c and 3d (early bubble of 'simple' keycodes) behave like 3a and 3b instead.

This is a preliminary merge, I would like to invite feedback on C coding style and conventions.

todo;
- [ ] rebase on BrainWarts' branch
- [ ] update docs

later
- [x] fix 3c/3d behavior
- [ ] add timing information to all keypress events, so we can handle nested tap-holds properly.

### CODE REVIEW

**Comment by okke-formsma** [comment]
The implementation requires a rather verbose behavior definition for every mod-tap. Instead, we could provide the user with a few predefined behaviors with that take 2 bindings, and a few macros to limit the amount of typing:

`MT(A, LCTL)`  instead of `&th-mt &kp A &kp LCTL`

`LT(A, 1)` instead of `&th-mt &kp A &mo 1`

File: app/include/zmk/event-manager.h [hunk 1]
```diff
@@ -78,6 +78,10 @@ struct zmk_event_subscription {
 #define ZMK_EVENT_RELEASE(ev) \
     zmk_event_manager_release((struct zmk_event_header *)ev);
 
+#define ZMK_EVENT_RELEASE_AGAIN(ev) \
+    zmk_event_manager_release_again((struct zmk_event_header *)ev);
```
Comment (okke-formsma) [inline app/include/zmk/event-manager.h:82]: the naming of this is not great, open to a better name.

File: app/src/behaviors/behavior_tap_hold.c [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+
+#if DT_NODE_EXISTS(DT_DRV_INST(0))
+
+/************************************************************  DATA SETUP */
+#define ZMK_BHV_TAP_HOLD_MAX_HELD 10
+#define ZMK_BHV_TAP_HOLD_MAX_CAPTURED_KC 40
+
+#define TH_TYPE_MOD_PREFERRED 0
+#define TH_TYPE_BALANCED 1
+#define TH_TYPE_TAP_PREFERRED 2
```
Comment (okke-formsma) [inline app/src/behaviors/behavior_tap_hold.c:32]: these 3  are not used anywhere

**Comment by BrainWart** [comment]
> MT(A, LCTL) instead of &th-mt &kp A &kp LCTL
So, formatting it like would be interesting. I think you would have to find a way to store how many binding cells each behavior would take. You may want to go the route of keeping the bindings property to hold the behaviors and use the cells associated with the behavior to store the data going to the other behaviors. Currently, the keymap supports a max of 2 cells that it will pass to the behavior.

```
mt: behavior-mod-tap {
    compatible = "zmk,behavior-tap-hold";
    label = "keypress keypress";
    #binding-cells = <2>;
    pass_params;
    tapping_term_ms = <300>;
    bindings = <&kp &kp>;
};
```

```
keymap {
    compatible = "zmk,keymap";
    label ="Default keymap";

    default_layer {
        bindings = < &mt LCTL A >;
    };
}
```

Passing `param1` to the first keypress behavior and `param2` to the second. Applying similar logic to a layer tap behavior if you wanted.

File: app/src/behaviors/behavior_tap_hold.c [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+  bool is_decided;
+  bool is_hold;
+  const struct behavior_tap_hold_config *config;
+  struct k_delayed_work work;
+  bool work_is_cancelled;
+};
+
+struct active_tap_hold* undecided_tap_hold = NULL;
+struct active_tap_hold active_tap_holds[ZMK_BHV_TAP_HOLD_MAX_HELD] = {};
+struct position_state_changed* captured_position_events[ZMK_BHV_TAP_HOLD_MAX_CAPTURED_KC] = {};
```
Comment (okke-formsma) [inline app/src/behaviors/behavior_tap_hold.c:63]: it's a little weird to use global variables here, but it's the only option I saw. The listener to key-position-events does not know for which device it was called. We need the globals to select the correct `active_tap_hold` struct.

File: app/src/behaviors/behavior_tap_hold.c [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+
+/************************************************************  CAPTURED POSITION HELPER FUNCTIONS */
+int capture_position_event(struct position_state_changed* event) {
```
Comment (okke-formsma) [inline app/src/behaviors/behavior_tap_hold.c:66]: should these be static?

File: app/src/behaviors/behavior_tap_hold.c [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+void decide_tap_hold(struct active_tap_hold * tap_hold, enum decision_moment event)
+{
+  if (tap_hold->is_decided) {
+    return;
+  }
+
+  if(tap_hold != undecided_tap_hold) {
+    LOG_DBG("ERROR found undecided tap hold that is not the active tap hold");
+    return;
+  }
```
Comment (okke-formsma) [inline app/src/behaviors/behavior_tap_hold.c:241]: this shouldn't happen ever, can we do something with asserts that are only built for debugging or something like that?

File: app/src/behaviors/behavior_tap_hold.c [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+  LOG_DBG("%d capturing %d %s event", undecided_tap_hold->position, ev->position, ev->state?"down":"up");
+
+  capture_position_event(ev);
+  if (ev->state) { 
+    decide_tap_hold(undecided_tap_hold, TH_OTHER_KEY_DOWN);
+  } else {
+    struct position_state_changed* captured_key_down = find_captured_keydown_event(ev->position); 
+    if(captured_key_down == NULL) {
+      // todo: allow key-up events for non-mod keys pressed before the TH was pressed.
+      // see scenario 3c/3d vs 3a/3b.
```
Comment (okke-formsma) [inline app/src/behaviors/behavior_tap_hold.c:367]: I wasn't sure how to implement this, as we're in a key position listener and don't know yet what kind of keypress will result here.

What is needed is something that finds the current behavior for a specific key position, so we can decide if it's a key press for a 'normal' key or something special like a momentary layer or a mod key.

File: app/src/event_manager.c [hunk 7]
```diff
@@ -24,24 +24,24 @@ int zmk_event_manager_handle_from(struct zmk_event_header *event, u8_t start_ind
...
-                        event->last_listener_index = i;
-                        // Listeners are expected to free events they capture
-                        return 0;
-                }
-            }
+            case ZMK_EV_EVENT_CAPTURED:
+                LOG_DBG("Listener captured the event");
+                event->last_listener_index = i;
+                // Listeners are expected to free events they capture
+                return 0;
```
Comment (okke-formsma) [inline app/src/event_manager.c:44]: this is a refactor that should not be in here. I'll take it out.

File: app/tests/tap-hold/.~lock.zmk-modtap-proposal.odg# [hunk 8]
```diff
@@ -0,0 +1 @@
+,okke,xps,30.08.2020 20:44,file:///home/okke/.config/libreoffice/4;
```
Comment (okke-formsma) [inline app/tests/tap-hold/.~lock.zmk-modtap-proposal.odg#:1]: oops, gotta get rid of this file too

**Comment by okke-formsma** [comment]
Closing this in favor of new PR for hold-taps.

**Pull request closed without merging**